rfd/0033-desktop-access.md

```diff
@@ -0,0 +1,176 @@
+
+Desktop Access works similarly to other existing Teleport services: an
+authenticated client application talks to the Teleport Proxy, which proxies the
+traffic to a backend service. Along the way, services enforce authorization and
+record the session.
+
+In Desktop Access, the client is a web application, bundled as part of the
+Teleport Web UI (similar to the web SSH client). The backend service (e.g.
+`windows_desktop_service`) is OS-specific and can represent multiple desktop
+hosts.
+
+```
++--------+
+| web UI |
++--------+
+    ^
+    | desktop protocol over websocket
+    v
++--------+
+| proxy  |
++--------+
+    ^
+    | desktop protocol over mTLS
+    v
++---------+
+| backend |
+| service |<----------------+
++---------+                 |
+   ^                        |
+   | OS protocol (e.g. RDP) |
+   v                        v
++---------+              +---------+
+| desktop |              | desktop |
+| host 1  |              | host 2  |
++---------+              +---------+
+```
+
+From web UI to the backend service, the underlying wire protocol is a custom
+protocol, which we'll call "desktop protocol". Between the backend service and
+the actual desktop hosts, the wire protocol is a standard protocol for the
+target OS (like RDP or X11), which we'll call "OS protocol".
+
+### Protocol
+
+The wire protocol used between the Teleport UI and backend service is a custom
+binary protocol. It is described in more detail in RFD 37.
+
+We are not using any standard remote desktop protocol, like
+RDP/X11/VNC/Guacamole for two reasons:
+- interoperability with 3rd party software is not a requirement
+- avoid legacy baggage and complexity of standard protocols, focus on
+  simplicity and performance
+
+We are also not using generic encoding protocols, like gRPC, to keep
+client-side parsing simple and reduce the JS bloat.
+
+### Client
+
+The client is a web app built into the Teleport Web UI. It's a JavaScript/TypeScript application that:
+- picks one of the allowed OS usernames for the logged-in Teleport user based
+  on their roles
+- creates a websocket connection
+- handles the incoming desktop protocol messages and renders them onto an HTML
+  canvas
+- captures user input, encodes it into desktop protocol messages and sends them
+  over the websocket
+
+TODO: wireframes?
+
+### Configuration
+
+There are no additional configuration options for Desktop Access in the web UI
+or the proxy. The Desktop Access client is part of the web UI and uses the same
+web port on the proxy for connecting a websocket.
+
+All relevant configuration to enable the feature will live in OS-specific
+backend services, described in their own RFDs (like RFD 34 for Windows).
+
+### Authorization
+
+Role definitions for RBAC have a few new fields for Desktop Access:
+- `${OS}_desktop_logins` - list of desktop login names allowed/denied for
+  desktop hosts with a given OS
+- `${OS}_desktop_labels` - list of labels to match the target desktop hosts
```

Comment:
How is matching between labels and actual hosts done?

Reply:
Same way as for SSH nodes and all other Teleport resources: https://goteleport.com/docs/access-controls/getting-started/

Resource labels are a set of `key: value` pairs.
Each Teleport role grants some set of corresponding `key: list of values` pairs.
When a user (via their roles) has a superset of all the `windows_desktop_labels` that a target desktop has, they are granted access.
When a resource has at least one label that the user doesn't have in their roles, access is denied.